core/src/main/java/org/jdbi/v3/core/argument/BuiltInArgumentFactory.java

```diff
@@ -13,218 +13,44 @@
-        Class<?> expectedClass = getErasedType(expectedType);
-
-        if (value != null && expectedClass == Object.class) {
-            expectedClass = value.getClass();
-        }
-
-        @SuppressWarnings("rawtypes")
-        ArgBuilder v = BUILDERS.get(expectedClass);
-
-        if (v != null) {
-            return Optional.of(v.build(value));
-        }
-
-        // Enums must be bound as VARCHAR.
-        if (value instanceof Enum) {
-            Enum<?> enumValue = (Enum<?>) value;
-            return Optional.of(STR_BUILDER.build(enumValue.name()));
-        }
-
-        if (value instanceof Optional) {
-            Object nestedValue = ((Optional<?>) value).orElse(null);
-            Type nestedType = findOptionalType(expectedType, nestedValue);
-            return config.get(Arguments.class).findFor(nestedType, nestedValue);
-        }
-
-        return value == null
-                ? Optional.of(config.get(Arguments.class).getUntypedNullArgument())
-                : Optional.empty();
-    }
-
-    private Type findOptionalType(Type wrapperType, Object nestedValue) {
-        if (getErasedType(wrapperType).equals(Optional.class)) {
-            Optional<Type> nestedType = findGenericParameter(wrapperType, Optional.class);
-            if (nestedType.isPresent()) {
-                return nestedType.get();
-            }
-        }
-        return nestedValue == null ? Object.class : nestedValue.getClass();
-    }
-
-    @FunctionalInterface
-    interface StatementBinder<T> {
-        void bind(PreparedStatement p, int index, T value) throws SQLException;
-    }
-
-    @FunctionalInterface
-    interface ArgBuilder<T> {
-        Argument build(T value);
-    }
-
-    static final class BuiltInArgument<T> implements Argument {
-        private final T value;
-        private final Class<T> klass;
-        private final int type;
-        private final StatementBinder<T> binder;
-
-        private BuiltInArgument(Class<T> klass, int type, StatementBinder<T> binder, T value) {
-            this.binder = binder;
-            this.klass = klass;
-            this.type = type;
-            this.value = value;
-        }
-
-        @Override
-        public void apply(int position, PreparedStatement statement, StatementContext ctx) throws SQLException {
-            if (value == null) {
-                statement.setNull(position, type);
-                return;
-            }
-            binder.bind(statement, position, value);
-        }
-
-        @Override
-        public String toString() {
-            if (klass.isArray()) {
-                return String.format("{array of %s length %s}", klass.getComponentType(), Array.getLength(value));
-            }
-            return String.valueOf(value);
-        }
-
-        StatementBinder<T> getStatementBinder() {
-            return binder;
-        }
+        return FACTORIES.stream()
+            .map(factory -> factory.build(expectedType, value, config))
```

Comment:
I prefer flatMap for this. It sucks they left out Optional.stream() in Java 8

```java
return FACTORIES.stream()
    .flatMap(factory -> JdbiStreams.toStream(
        factory.build(expectedType, value, config)))
    .findFirst();
```

Reply:
Is that really better though? I find it unintuitive to read with that custom stream tool in the middle.

Reply:
We could consider upping to Java 9 source level, which would give us access to `Optional.stream()`.  Java 8 is unsupported, after all... is it too soon?

Reply:
Java 8 is EOL already? I feel like it only just came out and the world is only just starting to use it by default.

Reply:
![image](https://user-images.githubusercontent.com/3950300/47117410-8a151480-d264-11e8-9a28-69b88e5850fb.png)

Reply:
AdoptOpenJDK will likely keep maintaining JDK 8 for a while after Oracle stops supporting it